qmlti/src/resources/LTI_Data_Connector_qmp.php

```diff
@@ -748,6 +748,66 @@ public function Results_getParticipantsByResource($consumer, $resource_link, $as
   }
 
 ###
+###  lTI_Attempts methods
+###
+
+/*
+ *  Gets the latest external attempt ID for an assessment
+ */
+  public function Attempts_getLatestAttempt($consumer_key, $resource_link_id, $assessment_id, $participant_id) {
+    $sql = 'SELECT schedule_id ' .
+           'FROM ' . $this->dbTableNamePrefix . LTI_Data_Connector::ATTEMPTS_TABLE_NAME . ' ' .
+           'WHERE (consumer_key = :consumer) AND (context_id = :context) AND (assessment_id = :assessment) AND (participant_id = :participant)';
+    $query = $this->db->prepare($sql);
+    $query->bindValue('consumer', $consumer_key, PDO::PARAM_STR);
+    $query->bindValue('context', $resource_link_id, PDO::PARAM_STR);
+    $query->bindValue('assessment', $assessment_id, PDO::PARAM_STR);
+    $query->bindValue('participant', $participant_id, PDO::PARAM_STR);
+    $ok = $query->execute();
+    if ($ok) {
+      $row = $query->fetch();
+      $schedule_id = $row['schedule_id'];
+    } else {
+      error_log(print_r($query->errorInfo(), 1));
+      return FALSE;
+    }
+    return $schedule_id;
+  }
+
+/*
+ *  Gets the latest external attempt ID for an assessment
+ */
+  public function Attempts_setLatestAttempt($consumer_key, $resource_link_id, $assessment_id, $schedule_id, $participant_id) {
+    $sql = 'INSERT INTO ' . $this->dbTableNamePrefix . LTI_Data_Connector::ATTEMPTS_TABLE_NAME .
+            ' (consumer_key, context_id, assessment_id, schedule_id, participant_id) ' .
+            'VALUES (:consumer, :context, :assessment, :schedule, :participant)';
+    $query = $this->db->prepare($sql);
+    $query->bindValue('consumer', $consumer_key, PDO::PARAM_STR);
+    $query->bindValue('context', $resource_link_id, PDO::PARAM_STR);
+    $query->bindValue('assessment', $assessment_id, PDO::PARAM_STR);
+    $query->bindValue('schedule', $schedule_id, PDO::PARAM_STR);
+    $query->bindValue('participant', $participant_id, PDO::PARAM_STR);
+    $ok = $query->execute();
+    return $ok;
+  }
+
+/*
+ *  Gets the latest external attempt ID for an assessment
```

Comment:
Missing PHPDoc. Probably a good idea if someone else has to maintain this in the future.